### ISSUE TITLE
Support for a feature gate flag/marker when generating CRDs

### ISSUE THREAD

**Issue by akutz** [kubernetes-sigs/controller-tools#600]
Has there ever been any consideration towards the inclusion of a feature gate flag / marker combination that could be used to control the generation of CRDs from the Go code? For example:

```go
type HelloWorldSpec struct {
}

type HelloWorldStatus struct {
}

// +kubebuilder:object:root=true
// +kubebuilder:subresource:status
// +kubebuilder:resource:scope=Namespace
// +kubebuilder:featuregate=MyFirstAPI
type HelloWorld struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec   HelloWorldSpec   `json:"spec,omitempty"`
    Status HelloWorldStatus `json:"status,omitempty"`
}
```

The presence of the marker `+kubebuilder:featuregate=MyFirstAPI` would indicate to the CRD generation that the type should not be considered unless `controller-gen` was executed with the following flag:

```bash
controller-gen \
  paths=hello-world/api/... \
  crd:trivialVersions=true \
  crd:crdVersions=v1 \
  crd:preserveUnknownFields=false \
  crd:featureGate=MyFirstAPI \
  output:crd:dir=./config/crd/bases \
  output:none
```

if a type does not specify a `kubebuilder:featuregate` annotation then it is not gated, but when a type *does* include such a marker, then this flag could prevent CRDs from being generated with features not yet ready for consumption.

There is a real business need for this because of the way private companies, even when developing in public, may want to align on API schemas in advance of releases. However, this presents a challenge because of deadlines and the risk of a feature not being ready. Still, the API was locked months ago. The flip side is organically updating an API over the duration of a release cycle. This is also not ideal because some API schema version will be constantly changing.

What would be nice is if new features could be gated at the API level behind a feature gate. That way the development could happen in the open, but the feature's API is only included in CRDs if:

* the feature gate is enabled when generating the CRDs -- useful for testing
* the feature gate is removed -- useful ahead of release

In looking into this I did come across `kubebuilder:skipversion`, but it does not seem to do *quite* the same thing.

Anyway, I'd love to hear others' thoughts and feedback on this. Thanks in advance!

**Comment by akutz** [comment]
cc @karaatanassov

**Comment by JoelSpeed** [comment]
For what it's worth, we have started using a patch in OpenShift when generating CRDs to have a distinction between fields with different feature sets (OpenShift groups feature gates into groups called feature sets).

We have multiple copies of the CRD, annotated with the desired feature set. The generator reads this annotation to identify if featuregeated fields should be included in the schema or not. OpenShift will then make sure the correct one is installed depending on which feature set the cluster is configured to use.

We haven't upstreamed this because it seemed like it was very specific to OpenShift and there are a number of questions it raises.
* How would you go backwards (ie what happens if you remove a field? - This doesn't matter in OpenShift as turning on a feature gate is a one way operation)
* How do users install their CRDs/how would they know which ones to install (In OpenShift this is operator driven, so we already have a solution)

I can see this being useful for the testing pipelines as already mentioned, but I think it does seem like the solution might need to be tailored quite specifically to each deployment system for Kubernetes, so I'm not sure how much value there is in an upstreamed solution to this

**Comment by sbueringer** [comment]
cc @JoelSpeed Given recent discussions

(not sure if you were aware (or remembered) that we have multiple issues for this, here is the other one: https://github.com/kubernetes-sigs/controller-tools/issues/1238)

**Comment by JoelSpeed** [comment]
For context, @sbueringer and I were discussing this morning that we probably want to have a proposal written up formally for this feature, since there are a lot of open questions and nuances around how feature gating would work/is expected to work for various folks.

@wazery and I have discussed a few in https://github.com/kubernetes-sigs/controller-tools/pull/1259

But things on my mind at the moment:
* We need a way to gate individual fields from being included/excluded depending on whether `MyFeature=true`/`MyFeature=false`
* I believe that the generator should fail if it finds a gate marker that it hasn't been given an enabled/disabled opinion for - to prevent accidents
* Do we promote people to use features as `MyCoolNewFeature` or levels of features such as `alpha`/`beta` (per @camilamacedo86's issue)
* If we go for features (as opposed to levels), what do we recommend to the community, to have multiple targets for generating their stable and then experimental channels? (I think that would align with the Gateway API project) 
* How do we want to tie in the upstream declarative validation approach to this, which will use `+k8s:enabledIf(MyFeatureGate)` and `+k8s:disabledIf(MyFeatureGate)` to control behaviour - we need a wider discussion about declarative validation and the future here as well
* Field enablement isn't complete, we also need to be able to gate other markers such as `enum` or `MaxItems` or `XValidation`
* We need a way to be able to pair gates with `AND` and `OR` semantics
* Given the complexity of this project, can we deliver it as an atomic unit or do we need to feature-gate this behaviour within controller-gen while we develop it?

This is a sufficiently complex area that I feel we cannot rush into this and should make sure we have consensus from both the maintainers and the community on what the best approach is to fit the needs here

**Comment by akutz** [comment]
> * We need a way to gate individual fields from being included/excluded depending on whether `MyFeature=true`/`MyFeature=false`

Yes! I updated the description to account for fields since I only mentioned them in a later comment.

I agree with most of your statements, but I also feel like this is a _real_ problem right now and it can be iterated upon without breaking anyone. If there was nothing but the ability to drop types/fields during generation based on the examples I provided, I feel like that is a really good starting point. To quote Joe, "Perfect is the enemy of good." It has been four years without a solution, and I really do not want to have to wait another four.

**Comment by akutz** [comment]
Either way, thank you again for picking up the baton on this @JoelSpeed . I really appreciate it!

**Comment by JoelSpeed** [comment]
Another thing to consider, is what is the correct behaviour for a feature gated field.

~In core K8s types, a feature gated field when not enabled, when sent by a client, will be rejected "this field is not enabled".~
Correction: In core K8s types, a feature gated field when not enabled, when sent by a client, it will be stripped before it gets to validation.

I believe in the examples where folks have gated fields in custom resources so far, a client who sends a gated when it is not enabled, gets it silently dropped.

~Should we actually be including the gated fields in the schemas, but with a reject message (likely implemented via CEL)?~

~What would the impact be of that change? For example, does it make changing feature gated fields stricter in terms of breaking changes?~

Omitting the fields from the schema is definitely the correct way to go, this comment can be ignored

**Comment by sbueringer** [comment]
> I believe in the examples where folks have gated fields in custom resources so far, a client who sends a gated when it is not enabled, gets it silently dropped.

I think this depends on what is configured for FieldValidation
```
	// FieldValidationIgnore ignores unknown/duplicate fields
	FieldValidationIgnore = "Ignore"
	// FieldValidationWarn responds with a warning, but successfully serve the request
	FieldValidationWarn = "Warn"
	// FieldValidationStrict fails the request on unknown/duplicate fields
	FieldValidationStrict = "Strict"
```
(pkg/apis/meta/v1)

I don't remember what the default is
